File: src/main/scala/bitlap/sbt/analyzer/parser/Graph.java
package bitlap.sbt.analyzer.parser;

import java.util.*;

public class Graph {
    private final int V;

    private final LinkedList<Integer>[] adj;

    @SuppressWarnings("unchecked")
    Graph(int v) {
        V = v;
        adj = new LinkedList[v];
        for (int i = 0; i < v; ++i)
            adj[i] = new LinkedList<>();
    }

    void addEdge(int v, int w) {
        adj[v].add(w);
    }

    void DFSUtil(int v, boolean[] visited, List<Integer> res) {
        visited[v] = true;

        res.add(v);

        for (int n : adj[v]) {
            if (!visited[n])
                DFSUtil(n, visited, res);
        }
    }

    public List<Integer> DFS(int v) {
        boolean[] visited = new boolean[V+1];
        List<Integer> res = new ArrayList<>();
        DFSUtil(v, visited, res);
        return res;
    }
}

File: src/main/scala/bitlap/sbt/analyzer/parser/Graph.scala
package bitlap.sbt.analyzer.parser;

import scala.collection.mutable.ListBuffer

final class Graph(size: Int) {

  private val graph = Array.fill(size)(ListBuffer[Int]())

  def addEdge(v: Int, w: Int): Unit = {
    graph(v) += w
  }

  def dfs(v: Int): ListBuffer[Int] = {
    val visited = Array.fill(size + 1)(false)
    val res     = ListBuffer[Int]()
    helper(v, visited, res)
    res
  }

  private def helper(v: Int, visited: Array[Boolean], res: ListBuffer[Int]): Unit = {
    visited(v) = true

    res += v

    graph(v).foreach(n => {
      if (!visited(n))
        helper(n, visited, res)
    })
  }
}
